### PR TITLE
Article timestamp coverage

### LINKED ISSUE #3838: Add missing ArticleTimestamp test coverage
**Is your feature request related to a problem? Please describe.**
The `ArticleTimestamp` component in Simorgh has very sparse unit test coverage and the business rules (esp in https://github.com/bbc/simorgh/blob/latest/src/app/containers/ArticleTimestamp/helpers.js
) are hard to follow and keep causing confusion.

**Describe the solution you'd like**
- Add missing unit test coverage

Then later we can refactor implementation code to make it more readable

**Describe alternatives you've considered**
None.

**Testing notes**
This should not involve any change in behaviour but manual regression testing is advisable. Additional unit test coverage should be added as part of this issue.

- [ ] This feature is expected to need manual testing.

**Additional context**

### PR DESCRIPTION
**Pull request by rhenshaw56**
Resolves #3838 

**Overall change:** adds unit test to cover logic in Article timestamp helpers
**Code changes:**

- adds unit test to cover logic in Article timestamp helpers

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] This PR requires manual testing

### CODE REVIEW

**Review by simonsinclair — APPROVED** [review]
Nice work. The tests cover `helpers.js`, are well described and all pass.

**Review by HarveyPeachey — APPROVED** [review]
LGTM

File: src/app/containers/ArticleTimestamp/helpers.test.js [hunk 1]
```diff
@@ -17,4 +26,94 @@ describe('ArticleTimestamp helper functions', () => {
       expect(isValidDateTime()).toEqual(false);
```
Comment (jamesdonoh) [inline src/app/containers/ArticleTimestamp/helpers.test.js]: There is no test for an `NaN` argument being passed, please could we add one?

File: src/app/containers/ArticleTimestamp/helpers.test.js [hunk 2]
```diff
@@ -17,4 +26,94 @@ describe('ArticleTimestamp helper functions', () => {
...
     });
   });
+
+  describe('isFirstRelative', () => {
+    it(`should return true when lastPublished === firstPublished and firstPublished < 10 hours ago`, () => {
+      const oneHourAgo = timestampGenerator({ hours: 1 });
+      const firstPublished = oneHourAgo;
+      const lastPublished = oneHourAgo;
+
+      expect(isFirstRelative(lastPublished, firstPublished)).toEqual(true);
```
Comment (jamesdonoh) [inline src/app/containers/ArticleTimestamp/helpers.test.js]: `toEqual` is supposed to be for comparing object properties (https://jestjs.io/docs/en/expect#toequalvalue), it may work but I think `toBe` is more appropriate in these tests?

File: src/app/containers/ArticleTimestamp/helpers.test.js [hunk 3]
```diff
@@ -17,4 +26,94 @@ describe('ArticleTimestamp helper functions', () => {
...
+    });
+
+    it(`should return false when lastPublished === firstPublished and firstPublished > 10 hours ago`, () => {
+      const fiveDaysAgo = timestampGenerator({ days: 5 });
+      const firstPublished = fiveDaysAgo;
+      const lastPublished = fiveDaysAgo;
```
Comment (jamesdonoh) [inline src/app/containers/ArticleTimestamp/helpers.test.js]: I think this also needs a test for when the two timestamps are unequal.

File: src/app/containers/ArticleTimestamp/helpers.test.js [hunk 4]
```diff
@@ -17,4 +26,94 @@ describe('ArticleTimestamp helper functions', () => {
...
+
+      expect(isFirstRelative(lastPublished, firstPublished)).toEqual(false);
+    });
+
+    it(`should return false when lastPublished !== firstPublished and firstPublished < 10 hours ago`, () => {
+      const firstPublished = timestampGenerator({ hours: 2 });
```
Comment (jamesdonoh) [inline src/app/containers/ArticleTimestamp/helpers.test.js:63]: Here and in the test for `isLastRelative` I would look at testing a few more boundary cases. I could change the implementation to use an imaginary `isEightHoursAgo` function instead of `isTenHoursAgo` and these would still pass.

File: src/app/containers/ArticleTimestamp/helpers.test.js [hunk 5]
```diff
@@ -17,4 +26,94 @@ describe('ArticleTimestamp helper functions', () => {
...
+    });
+
+    it('should return false when lastPublished > 10 hours ago', () => {
+      const lastPublished = timestampGenerator({ days: 1 });
+      expect(isLastRelative(lastPublished)).toEqual(false);
+    });
+  });
+
+  describe('isSameDay', () => {
+    it('should return true if both timestamps are within 24 hours of each other', () => {
```
Comment (jamesdonoh) [inline src/app/containers/ArticleTimestamp/helpers.test.js]: Are the descriptions here and L74 correct? I thought the logic checked if they are the same calendar day, not that they are in the same 24h window. Or am I mistaken?

File: src/app/containers/ArticleTimestamp/helpers.test.js [hunk 6]
```diff
@@ -17,4 +26,94 @@ describe('ArticleTimestamp helper functions', () => {
...
+      expect(isToday(timestamp)).toEqual(false);
+    });
+  });
+
+  describe('formatType', () => {
+    const dateFormats = {
+      date: 'D MMMM YYYY',
+      dateTimeTimezone: 'D MMMM YYYY, HH:mm z',
+    };
+    it(`should return default date format when firstPublished is > 10 hours ago`, () => {
```
Comment (jamesdonoh) [inline src/app/containers/ArticleTimestamp/helpers.test.js]: Hmm, this seems a little odd to me - the implementation only uses `isSameDay` and `isToday`, so why why the figure of 10 hours significant here?

This function is the most complicated and difficult one to read in the whole file IMO. So we need to make sure the expected output for every significant combination of arguments are tested, otherwise we will be unable to detect certain kinds of regression. One way of doing this is identifying every condition mentioned in the implementation, e.g.:

```
isToday(lastPublished) 
isToday(firstPublished)
lastPublished (truthiness)
firstPublished (truthiness)
isSameDay(lastPublished, firstPublished)
```
Don't forget to consider negation too (e.g. what if `firstPublished` is falsy/null?)

**Review by jamesdonoh — CHANGES_REQUESTED** [review]

File: src/app/containers/ArticleTimestamp/helpers.test.js (hunk 6 above)
Comment (rhenshaw56) [reply src/app/containers/ArticleTimestamp/helpers.test.js]: makes sense

File: src/app/containers/ArticleTimestamp/helpers.test.js [hunk 7]
```diff
@@ -1,20 +1,137 @@
...
+      expect(isValidDateTime(null)).toBe(false);
+      expect(isValidDateTime(undefined)).toBe(false);
+      expect(isValidDateTime()).toBe(false);
+      expect(isValidDateTime(NaN)).toBe(false);
+    });
+  });
+
+  describe('isFirstRelative', () => {
+    describe(`when lastPublished === firstPublished`, () => {
+      const timeLimits = generateTimeLimits(15);
```
Comment (jamesdonoh) [inline src/app/containers/ArticleTimestamp/helpers.test.js]: The helper is a good idea, and the output is very readable. But I think doing 1-15h is probably overkill and produces too much output. Let's just test 1h, 9h and 10h as the most important boundaries.

File: src/app/containers/ArticleTimestamp/helpers.test.js [hunk 8]
```diff
@@ -1,20 +1,137 @@
...
+      expect(isToday(timestamp)).toBe(false);
+    });
+  });
+
+  describe('formatType', () => {
+    const dateFormats = {
+      date: 'D MMMM YYYY',
+      dateTimeTimezone: 'D MMMM YYYY, HH:mm z',
+    };
+    it(`should return default date format when firstPublished is not today`, () => {
```
Comment (jamesdonoh) [inline src/app/containers/ArticleTimestamp/helpers.test.js]: I'd remove the word 'default' here and in the tests below as there isn't really what you could accurately call a default case in this function, e.g. "return date format"

File: src/app/containers/ArticleTimestamp/helpers.test.js [hunk 9]
```diff
@@ -1,20 +1,137 @@
...
+      const firstPublished = timestampGenerator({ days: 5 });
+      expect(formatType({ firstPublished })).toBe(dateFormats.date);
+    });
+
+    it(`should return dateTimeTimezone format when firstPublished is today`, () => {
```
Comment (jamesdonoh) [inline src/app/containers/ArticleTimestamp/helpers.test.js]: "dateTimeTimezone" has no meaning outside these tests, could we say "date and time with timezone format"? (Here and below)

File: src/app/containers/ArticleTimestamp/helpers.test.js [hunk 10]
```diff
@@ -1,20 +1,137 @@
...
+      const firstPublished = timestampGenerator({ hours: 4 });
+      const lastPublished = timestampGenerator({ hours: 2 });
+      expect(formatType({ firstPublished, lastPublished })).toBe(
+        dateFormats.dateTimeTimezone,
+      );
+    });
+
+    it(`should return default date format when firstPublished is null`, () => {
+      const firstPublished = null;
+      expect(formatType({ firstPublished })).toBe(dateFormats.date);
```
Comment (jamesdonoh) [inline src/app/containers/ArticleTimestamp/helpers.test.js:137]: I don't think this test correctly reflects the logic in the implementation due to the way default parameters work. Here we are passing an explicit `null` to `formatType` as the first argument (lastPublished), not setting firstPublished to null. And no I'm not sure why formatDate was implemented that way :-/

**Review by jamesdonoh — CHANGES_REQUESTED** [review]

File: src/app/containers/ArticleTimestamp/helpers.test.js (hunk 10 above)
Comment (rhenshaw56) [reply src/app/containers/ArticleTimestamp/helpers.test.js:137]: @jamesdonoh `lastPublished` isn't really the first argument as `formatType` takes in an object, `firstPublished` will be null and `lastPublished` will default to `null`

Comment (jamesdonoh) [reply src/app/containers/ArticleTimestamp/helpers.test.js:137]: You're absolutely right, apologies. I missed the curly brackets :-)

File: src/app/containers/ArticleTimestamp/helpers.test.js [hunk 11]
```diff
@@ -1,20 +1,140 @@
...
+    const dateFormats = {
+      date: 'D MMMM YYYY',
+      dateTimeTimezone: 'D MMMM YYYY, HH:mm z',
+    };
+    it(`should return date format when firstPublished is not today`, () => {
+      const firstPublished = timestampGenerator({ days: 5 });
+      expect(formatType({ firstPublished })).toBe(dateFormats.date);
+    });
+
+    it(`should return date and time with timezone format format when firstPublished is today`, () => {
```
Comment (jamesdonoh) [inline src/app/containers/ArticleTimestamp/helpers.test.js]: typo `format format` here and below

**Review by jamesdonoh — APPROVED** [review]
Looks good apart from typo, thanks for the changes. 👍
